src/path.cr

```diff
@@ -253,24 +253,30 @@ struct Path
   # ```
   def each_parent(&block : Path ->)
     return if @name.empty? || @name == "."
-
+    reader = Char::Reader.new(@name)
+    last_was_separator = false
+    last_sep_pos = 0
     first = true
-    each_part_separator_index do |pos|
```

Comment:
I think we should keep using `each_part_separator_index`. It's just horribly broken and the issues need to be fixed *there*.
`Path#part` uses the same helper and suffers from the same issues.

It seems there are also no specs for this method. Here are a few:
```cr

  assert_paths_raw("m/.gitignore", ["m", ".gitignore"], &.parts)
  assert_paths_raw("m", ["m"], &.parts)
  assert_paths_raw("m/", ["m"], &.parts)
  assert_paths_raw("m//", ["m\\"], ["m"], &.parts)
  assert_paths_raw("m//a/b", ["m\\a", "b"], ["m", "a", "b"], &.parts)
  assert_paths_raw("/m", ["/", "m"], &.parts)
  assert_paths_raw("/m/", ["/", "m"], &.parts)
  assert_paths_raw("C:", ["C:"], &.parts)
  assert_paths_raw("C:/", ["C:/"], &.parts)
  assert_paths_raw("C:\\", ["C:\\"], &.parts)
  assert_paths_raw("C:folder", ["C:folder"], ["C:", "folder"], &.parts)
  assert_paths_raw("C:\\folder", ["C:\\folder"], ["C:\\", "folder"], &.parts)
  assert_paths_raw("C:\\\\folder", ["C:\\\\folder"], ["C:\\", "folder"], &.parts)
  assert_paths_raw("C:\\.", ["C:\\."], ["C:\\"], &.parts)
```

Alternatively, instead of duplicating the expectations for `parents` and `parts`, we could just combine them.
The following should always be true:
```cr
parts == parents.map(&.basename) + basename

# The reverse is only valid for normalized paths because `#parts` loses information
parents == parts[1..].reduce([Path.new(parts.first)]) do |paths, part|
  paths << paths.last.join(part)
end
```

Reply:
I don't think `each_part_separator_index` is broken, i just see it impractical for implementation of `each_parent`. I like to yield on other char than separator. I try with `each_part_separator_index` and made 3 very messy and confused implementations.. Maybe you have some elegant variant in mind?

Reply:
I also chose a different approach to implement `parts` which seems very simple and clear for me. See last commit.

Reply:
The very purpose of `each_part_separator_index` is to provide the basic behaviour that is similar to `#parts` and `#parents` (and `#anchor` uses it as well). So if we couldn't use it in `#parts`, it wouldn't make much sense to have such a helper method.